spock-core/src/main/java/org/spockframework/mock/runtime/IMockMaker.java

```diff
@@ -0,0 +1,317 @@
+package org.spockframework.mock.runtime;
+
+import org.spockframework.mock.CannotCreateMockException;
+import org.spockframework.mock.IMockObject;
+import org.spockframework.mock.MockNature;
+import org.spockframework.util.Beta;
+import org.spockframework.util.Immutable;
+import org.spockframework.util.Nullable;
+import org.spockframework.util.ThreadSafe;
+import spock.mock.MockMakers;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * The {@link IMockMaker} provides a {@link java.util.ServiceLoader} extension point to allow the usage
+ * of different implementations to create the underlying mock classes and mock instances.
+ *
+ * <p>
+ * The class {@link MockMakers} lists the well known {@link IMockMaker} currently defined in Spock.
+ *
+ * <p> You can override the used {@link IMockMaker} with the {@link MockMakerRegistryConfiguration}.
+ * Please note that the {@link MockMakers#javaProxy} has always priority over every other {@link IMockMaker}.
+ *
```

Comment:
I do not really get that with only the first should be a link. 
What is the rational behind that? This make the editing in the IDE more cumbersome, and the `{@code Type}` does not find the declaration in the IDE, also the content is not checked for wrong type links in `{@code }`. 
Also the JRE-Classes do not do that kind of thing, e.g. `CompletableFuture` => `CompletionStage`
And also Spock classes don't do that consistently, e.g.  `TempDir` 

Also why do you want in some places the structure `{@link IMockCreationSettings IMockCreationSettings}`?
This would be the same as `{@link IMockCreationSettings}` and the long version could contain typos.

Reply:
> I do not really get that with only the first should be a link. What is the rational behind that?

That is to improve readability. Each link draws reader attention to itself by coloring / underlining. Also think about visually impaired developers where the braille reader or the screen reader needs to tell it is a link on every other word.

It is common practice to not link where not necessary and where necessary just the first occurence.

It is also part of the JavaDoc Style Guide at https://www.oracle.com/de/technical-resources/articles/java/javadoc-tool.html:
> Use in-line links economically
>
> You are encouraged to add links for API names (listed immediately above) using the {@link} tag. It is not necessary to add links for all API names in a doc comment. Because links call attention to themselves (by their color and underline in HTML, and by their length in source code doc comments), it can make the comments more difficult to read if used profusely. We therefore recommend adding a link to an API name if:
>
> - The user might actually want to click on it for more information (in your judgment), and
> - Only for the first occurrence of each API name in the doc comment (don't bother repeating a link)
>
> Our audience is advanced (not novice) programmers, so it is generally not necessary to link to API in the java.lang package (such as String), or other API you feel would be well-known.

> Also the JRE-Classes do not do that kind of thing, e.g. CompletableFuture => CompletionStage 

The JRE classes usually follow that style guide.
In `CompletableFuture` class doc there is 6 times `CompletionStage`, but it is only linked 2 times.
The second link is probably more a documentation bug then. :-)
In some of the method comments it even is not a single time linked.

> And also Spock classes don't do that consistently, e.g. TempDir

Noone said the Spock codebase is consistent or clean. :-D
For such an old and big codebase where many people had their hands on you always have some inconsistencies.
Doesn't mean we cannot improve for the future. :-)

> Also why do you want in some places the structure {@link IMockCreationSettings IMockCreationSettings}? This would be the same as {@link IMockCreationSettings} and the long version could contain typos.

Just a suggestion for improved readability:
![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/tdLND5eJpRKUT3xAEJML/fe7f014c-21c9-4be5-9fcd-e89bd90e7dce/image.png)
![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/tdLND5eJpRKUT3xAEJML/8e880d02-c408-4968-8366-8b699ce4e3a7/image.png)